eppo_client/bandit.py

```diff
@@ -0,0 +1,267 @@
+    action_attributes: Optional[Attributes]
+    action_score: float
+    action_weight: float
+    gamma: float
+
+
+@dataclass
+class BanditResult:
+    variation: str
+    action: Optional[str]
+
+
+def null_evaluation(
+    flag_key: str, subject_key: str, subject_attributes: Attributes, gamma: float
+):
+    return BanditEvaluation(
+        flag_key,
+        subject_key,
+        subject_attributes,
+        None,
+        None,
+        0.0,
+        0.0,
+        gamma,
+    )
+
+
+@dataclass
+class BanditEvaluator:
+    sharder: Sharder
+    total_shards: int = 10_000
+
+    def evaluate_bandit(
+        self,
+        flag_key: str,
+        subject_key: str,
+        subject_attributes: Attributes,
+        actions_with_contexts: List[ActionContext],
+        bandit_model: BanditModelData,
+    ) -> BanditEvaluation:
+        # handle the edge case that there are no actions
+        if not actions_with_contexts:
+            return null_evaluation(
+                flag_key, subject_key, subject_attributes, bandit_model.gamma
+            )
+
+        action_scores = self.score_actions(
+            subject_attributes, actions_with_contexts, bandit_model
+        )
+
+        action_weights = self.weigh_actions(
+            action_scores,
+            bandit_model.gamma,
+            bandit_model.action_probability_floor,
+        )
+
+        selected_idx, selected_action = self.select_action(
+            flag_key, subject_key, action_weights
+        )
+        return BanditEvaluation(
+            flag_key,
+            subject_key,
+            subject_attributes,
+            selected_action,
+            actions_with_contexts[selected_idx].attributes,
+            action_scores[selected_idx][1],
+            action_weights[selected_idx][1],
+            bandit_model.gamma,
+        )
+
+    def score_actions(
+        self,
+        subject_attributes: Attributes,
+        actions_with_contexts: List[ActionContext],
+        bandit_model: BanditModelData,
+    ) -> List[Tuple[str, float]]:
+        return [
+            (
+                action_context.action_key,
+                (
+                    score_action(
+                        subject_attributes,
+                        action_context.attributes,
+                        bandit_model.coefficients[action_context.action_key],
+                    )
+                    if action_context.action_key in bandit_model.coefficients
+                    else bandit_model.default_action_score
+                ),
+            )
+            for action_context in actions_with_contexts
+        ]
+
+    def weigh_actions(
+        self, action_scores, gamma, probability_floor
+    ) -> List[Tuple[str, float]]:
+        number_of_actions = len(action_scores)
+        best_action, best_score = max(action_scores, key=lambda t: t[1])
+
+        # adjust probability floor for number of actions to control the sum
+        min_probability = probability_floor / number_of_actions
```

Comment:
The way it's coded in Java is the probability floor is an absolute floor irrespective of the number of actions ([src](https://github.com/Eppo-exp/java-server-sdk/blob/main/src/main/java/com/eppo/sdk/helpers/bandit/FalconBanditModel.java#L82)). Either way, it has its issues, but dividing it by the number of actions seems safer if you'd like that to become the standard.

Reply:
Yup I think this is safer; otherwise there is no good way to set the probability floor generally. Suppose 1 bandit gets called with 1000s of actions, and the other one with 5; what probability floor should we be using if we don't normalize?

Reply:
💯